packages/core/src/components/index.ts

```diff
@@ -4,14 +4,6 @@
  * Licensed under the terms of the LICENSE file distributed with this project.
  */
 
-declare function require(moduleName: string): any; // declare node.js "require" so that we can conditionally import
-if (typeof window !== "undefined" && typeof document !== "undefined") {
-    // we're in browser
-    // tslint:disable-next-line:no-var-requires
-    require("dom4"); // only import actual dom4 if we're in the browser (not server-compatible)
```

Comment:
unfortunately, this is technically a breaking change. this is a side-effectful import which allows blueprint users to use dom4 features without explicitly importing the polyfill themselves, so removing it can cause runtime errors

Reply:
I would recommend implementing the ponyfill (as you've done in this PR) but leaving this import line in for now... you can prevent "dom4" from loading in your webpack config... and rip the band-aid off in the next major version

Reply:
> is removing a side effect a breaking change?

For regular JS (un-typechecked) usage, yes, I think it's a breaking change. BP users are currently getting dom4 pulled in transitively and they are able to use all those features in their code (possibly without realizing they are). Removing this dependency in a minor version can cause runtime breaks. _However_, TS consumers will get compilation errors if they are using those features and `node_modules/@types/dom4` is removed as a result of this change. So it's a less dangerous break for typechecked usage. Still, I would err on the side of caution and try not to introduce build failures in a minor version bump.

Also, it's not a dynamic import, those use the `import()` syntax.

> also does not make sense to leave the import if we remove the dependency from package.json.

yeah, I agree, so you'd need to keep the dependency in package.json (remove `@types/dom4` though). for our BP1 + BP3 scenario, as I suggested above, we can prevent dom4 from getting on the page through webpack config (and we can recommend that other users do this too, if they don't want the dead module import pulled into their bundle).

Reply:
these dom4 methods are not a part of blueprint's public API or our environment requirements (because we provide them for ourselves). so if a user is relying on a dom4 method, i'd argue it's actually on _them_ to ensure that their environment provides it, rather than relying on the transitive inclusion here in blueprint.

Reply:
> i'd argue it's actually on them to ensure that their environment provides it

sure, but we haven't made that explicit in our documentation or anything, there wasn't any contract there, so it still feels like a break to me. the sticking point for me is the fact that it's easy to use the dom4 methods on DOM primitives without thinking about the browser support (perhaps you're copy-pasting from the BP source, etc). is it a huge deal to leave the import in here, given the workaround I proposed?

Reply:
for what it's worth, a similar discussion happened with the removal of `es6-shim` from `@blueprintjs/table` if prior art would help - https://github.com/palantir/blueprint/issues/522